src/main/scala/uk/gov/nationalarchives/omega/api/repository/RepositoryUtils.scala

```diff
@@ -103,7 +103,8 @@ trait RepositoryUtils extends AgentTypeMapper {
       queryText_2 <- Try(setResourceParams(queryText_1, params.uris))
```

Comment:
I wonder if there is a better syntax for this perhaps. It feels to me like two different things but not quite either:
1. a Zip
2. a Left Fold.

Reply:
Or perhaps better yet to avoid creation/gc of many mutable `ParameterizedSparqlString` objects... we could do something like this:

```scala
  private def parameterizeQuery(queryText: String, params: SparqlParams): Try[ParameterizedSparqlString] = {
    val parameterize = Seq(
      setBooleanParams(_, params.booleans),
      setResourceParams(_, params.uris),
      setValueParams(_, params.values),
      setXSDDateTimeParams(_, params.dateTimes),
    ).reduceLeft(_.andThen(_))

    val filteredQueryText = setFilterParams(queryText, params.filters)
    Try(parameterize(new ParameterizedSparqlString(filteredQueryText)))
  }

  private def setParams[T](query: ParameterizedSparqlString, params: Map[String, T], set: (ParameterizedSparqlString, T) => String => Unit): ParameterizedSparqlString = {
    params.foreach(param => set(query, param._2)(param._1))
    query
  }

  private def setBooleanParams(query: ParameterizedSparqlString, booleanParams: Map[String, Boolean]): ParameterizedSparqlString =
    setParams[Boolean](query, booleanParams, (query, booleanParamValue) => query.setLiteral(_: String, booleanParamValue))

  private val dateTimeType = new XSDDateTimeType("dateTime");

  private def setXSDDateTimeParams(query: ParameterizedSparqlString, dateTimeParams: Map[String, XMLGregorianCalendar]): ParameterizedSparqlString =
    setParams[XMLGregorianCalendar](query, dateTimeParams, (query, dateTimeParamValue) => query.setLiteral(_: String, dateTimeParamValue.toXMLFormat, dateTimeType))

  private def setResourceParams(query: ParameterizedSparqlString, resourceParams: Map[String, String]): ParameterizedSparqlString =
    setParams[String](query, resourceParams, (query, resourceParamValue) => query.setIri(_: String, resourceParamValue))

  private def setValueParams(query: ParameterizedSparqlString, valueParams: Map[String, List[Resource]]): ParameterizedSparqlString =
    setParams[List[Resource]](query, valueParams, (query, valueParamValue) => query.setValues(_: String, valueParamValue.asJava))

  private def setFilterParams(queryText: String, filterParams: Map[String, String]): String = {
    @tailrec
    def setParams(filterParams: Map[String, String], query: String): String = {
      filterParams match {
        case m: Map[String, String] if m.isEmpty => query
        case param =>
          val queryWithFilter = query.replace(s"?${param.head._1}", param.head._2)
          setParams(filterParams.tail, queryWithFilter)
      }
    }
    setParams(filterParams, queryText)
  }
```

Reply:
I've made the change you suggested and it works fine. It took me a little while to figure out how the `parameterize` function is working so I added the type, just to make it a bit more obvious.